trillian/integration/ct_integration.go

```diff
@@ -437,80 +501,63 @@ func RunCTIntegrationForLog(cfg *configpb.LogConfig, servers, metricsServers, te
-		return fmt.Errorf("len(entries)=%d; want %d", len(precertEntries), 1)
-	}
-	leaf := precertEntries[0].Leaf
-	ts := leaf.TimestampedEntry
-	fmt.Printf("%s: Entry[%d] = {Index:%d Leaf:{Version:%v TS:{EntryType:%v Timestamp:%v}}}\n",
-		t.prefix, precertIndex, precertEntries[0].Index, leaf.Version, ts.EntryType, timeFromMS(ts.Timestamp))
-
-	if ts.EntryType != ct.PrecertLogEntryType {
-		return fmt.Errorf("leaf[%d].ts.EntryType=%v; want PrecertLogEntryType", precertIndex, ts.EntryType)
-	}
-	if !bytes.Equal(ts.PrecertEntry.TBSCertificate, tbs) {
-		return fmt.Errorf("leaf[%d].ts.PrecertEntry differs from originally uploaded cert", precertIndex)
-	}
-	if err := t.checkStats(); err != nil {
-		return fmt.Errorf("unexpected stats check: %v", err)
+	if err := t.checkPreCertEntry(ctx, precertIndex, tbs); err != nil {
+		return fmt.Errorf("failed to check pre-cert entry: %v", err)
 	}
 
 	// Stage 14: get an inclusion proof for the precert.
-	// Calculate leaf hash =  SHA256(0x00 | tls-encode(MerkleTreeLeaf))
-	leaf = ct.MerkleTreeLeaf{
-		Version:  ct.V1,
-		LeafType: ct.TimestampedEntryLeafType,
-		TimestampedEntry: &ct.TimestampedEntry{
-			Timestamp: precertSCT.Timestamp,
-			EntryType: ct.PrecertLogEntryType,
-			PrecertEntry: &ct.PreCert{
-				IssuerKeyHash:  sha256.Sum256(issuer.RawSubjectPublicKeyInfo),
-				TBSCertificate: tbs,
-			},
-			Extensions: precertSCT.Extensions,
-		},
-	}
-	leafData, err := tls.Marshal(leaf)
-	if err != nil {
-		return fmt.Errorf("tls.Marshal(precertLeaf)=(nil,%v); want (_,nil)", err)
-	}
-	hash := sha256.Sum256(append([]byte{merkletree.LeafPrefix}, leafData...))
-	rsp, err := t.client().GetProofByHash(ctx, hash[:], sthN1.TreeSize)
-	t.stats.done(ctfe.GetProofByHashName, 200)
-	if err != nil {
-		return fmt.Errorf("got GetProofByHash(precertSCT, size=%d)=nil,%v", sthN1.TreeSize, err)
-	}
-	if rsp.LeafIndex != precertIndex {
-		return fmt.Errorf("got GetProofByHash(precertSCT, size=%d).LeafIndex=%d; want %d", sthN1.TreeSize, rsp.LeafIndex, precertIndex)
-	}
-	fmt.Printf("%s: Inclusion proof leaf %d @ %d -> root %d = %x\n", t.prefix, precertIndex, precertSCT.Timestamp, sthN1.TreeSize, rsp.AuditPath)
-	if err := Verifier.VerifyInclusionProof(rsp.LeafIndex, int64(sthN1.TreeSize), rsp.AuditPath, sthN1.SHA256RootHash[:], leafData); err != nil {
-		return fmt.Errorf("got VerifyInclusionProof(%d,%d,...)=%v; want nil", precertIndex, sthN1.TreeSize, err)
-	}
-	if err := t.checkStats(); err != nil {
-		return fmt.Errorf("unexpected stats check: %v", err)
+	if err := t.checkInclusionOfPreCert(ctx, tbs, issuer, precertSCT, sthN1); err != nil {
+		return fmt.Errorf("failed to check inclusion of pre-cert entry: %v", err)
 	}
 
 	// Stage 15: invalid consistency proof
 	if rsp, err := t.client().GetSTHConsistency(ctx, 2, 299); err == nil {
 		return fmt.Errorf("got GetSTHConsistency(2,299)=(%+v,nil); want (nil,_)", rsp)
 	}
-	t.stats.done(ctfe.GetSTHConsistencyName, 400)
+	t.stats.done(ctfe.GetSTHConsistencyName, 500)
```

Comment:
Feels like these should be `400`s for duff data, maybe there's a bug elsewhere :)